### DISCUSSION TITLE
Provide handler in test that interacts with server-side NextJS code

### DISCUSSION THREAD

**Discussion by chrisb2244** [mswjs/msw#1431] [Q&A]
I'd like to make assertions about the behaviour of my NextJS application, with varying levels of mocking.

As a concrete example, I have an API route that interacts with a database to check if a user exists (I can mock with MSW, or allow this to actually run against the remote DB) and then send an invitation email or signin magic link or similar (via an email service, which I'd like to mock, and get the generated link and text for in my tests).

An alternative is to use a service to have an API for a mailbox, and send real emails, but I'd prefer not to do this.

The server can use MSW with some code like the following, to avoid sending real emails:
```
// handlers.ts
import { rest } from 'msw'

export const handlers = [
 rest.post('https://api.mailgun.net/*/messages', (req, res, ctx) => {
    // don't send email, just assert that it worked
    return res(ctx.status(200))
  })
]
```

```
// server.ts
import { handlers } from './handlers'

const requestInterceptorFactory = (() => {
    const { setupServer } = require('msw/node') as typeof import('msw/node')
    return setupServer(...handlers)
})

export const server = requestInterceptorFactory()
```

```
// mocks/index.ts
export const initMocks = async () => {
  if (typeof window === 'undefined') {
    const { server } = await require('./server') as typeof import('./server')
    server.listen({
      onUnhandledRequest: 'bypass'
    })
    return server
  } else {
    const { worker } = await import('./browser')
    worker.start()
    return worker
  }
}

```

```
// _app.tsx
if (process.env.PLAYWRIGHT === '1') {
  import('../../playwright/mocks').then(({initMocks}) => {
    initMocks()
  })
}

const MyApp: React.FC<{blah: Blah}> = (props) => {
  return (
      // application goes here
  )
}

export default MyApp
```

but this doesn't allow me to check the value of the data that was sent to the endpoint (which I could get from `req.json()` or similar).

Adding something like
```
export const requestInterceptor = (process.env.PLAYWRIGHT === '1' && typeof window === 'undefined')
  ? (() => {
    console.log('Setting up reqHandler on server in app')
    const { setupServer } = require('msw/node')
    const requestInterceptor: SetupServerApi = setupServer()
    requestInterceptor.listen()
    return requestInterceptor
  }) ()
  : undefined
```
to the `_app.tsx` seems to be in line with the process described here: https://frontend-digest.com/using-playwright-to-test-next-js-applications-80a767540091 (but here without building - I also couldn't seem to make that work).

I can then in a fixture import via code like:
```
import { test as base } from '@playwright/test'
import { rest } from 'msw'
import type { SetupServerApi } from 'msw/node'

export const test = base.extend<{
  requestInterceptor: SetupServerApi
  rest: typeof rest
}>({
  requestInterceptor: [
    async ({}, use) => {
      // This seems required due to some middleware typing :/
      const { Request, Response } = require('node-fetch')
      global.Request = Request
      global.Response = Response

      // Here 'dummy-next' is a Docker volume for the dev copy's ".next" directory,
      // you can also use "../.next/server/pages/_app" to get the local version if running locally
      const { requestInterceptor } = await require('../dummy-next/server/pages/_app')
      // console.log(requestInterceptor)
      use(requestInterceptor)
    }, {
      //@ts-ignore
      scope: 'test', // test?
    }
  ],
  rest,
})
```

Here (so long as I `await` the require), requestInterceptor is defined and seems sensible, but doesn't trigger even if I add handlers in the test:
```
// example-test.ts
import { test, expect } from './next-fixture'
test('User can log-in', async ({page, rest, requestInterceptor}) => {
  console.group('Debugging for reqI')
  console.log('reqI: ', requestInterceptor)
  requestInterceptor.printHandlers() 
  // prints handlers defined in the setupServer call within the requestInterceptor defining function in _app.tsx
  console.groupEnd()

  requestInterceptor.use(
    rest.all('*', (req) => {
      // This never triggers
      console.log('URL: ', req.url, '\nMethod: ', req.method)
      return req.passthrough()
    })
  )

  // do stuff
```

Is this possible? Can the test instance gain access to the server-side code?

I managed to assert against the application server's behaviour using an http server and custom headers as described here: https://github.com/mswjs/msw/discussions/1322#discussioncomment-3836634 but that requires test code in the real application, and isn't ideal.

**Comment by kettanaito** [comment]
Hey, @chrisb2244. That's a good question. 

The thing is, your test process (Playwright) and the Next server process are two different processes. You cannot directly access any data in the server process unless there's some sort of communication channel implemented explicitly by both parties. 

To make things even more interesting, the Next server is _not_ a child process of `next run` (or whichever command is used to run Next). I was testing it in the past and that was not the case. I have no idea how to even reference that server process to begin with. This is something the Next team may have a better answer for. 

I don't think you can access the intercepted request info from your handler that runs in a separate server process from the Playwright test process. We've started a remote interception experiment (#799), and while it's successful for parent <-> child processes, the Next server is not a child of the spawned `next run` process, as I've mentioned. At the moment, this is not supported. 

## Suggestion: Assert the response

We generally recommend against request assertions. So, whenever you can, try writing assertions around the response that your app renders. If that's not applicable, only then resort to request-based assertions.

I think this may be your case, as sending an email request is likely a side-effect that may not have any observable state in your app.

**Reply by chrisb2244** [reply]
Hey @kettanaito, thanks for the response.

I saw the branch you mentioned (https://github.com/mswjs/msw/pull/799) and also the linked PR (https://github.com/mswjs/interceptors/pull/129) although when I tried to find them again, I couldn't remember where - so thank you also for the links! Those helped me understand some of the problems with what I really wanted, and guided me to looking for an 'out-of-band' method, so to say.

I did manage to replace my code in the linked thread I originally mentioned (https://github.com/mswjs/msw/discussions/1322#discussioncomment-3836634) using only MSW by passing information in the "email" field, which is received in both of my endpoints for this particular case (and also has the benefit of being available for the 'signed in user', in other handlers if required).
This still leaves me with only compile-time (or whatever the equivalent term is) handlers for the server-side, but potentially by adding more information into the email address, I could `switch` behaviours (ugly, but workable).

```
// example handler
import { rest } from 'msw'

const emailToEndpoint = (email: string) => {
  const attemptedEndpoint = email.match(/test[0-9]+_([0-9]+\.[0-9]+\.[0-9]+\.[0-9]+)_([0-9]+)_(.*)_user@specificemailaddress.com/)
  if (attemptedEndpoint !== null) {
    const ip = attemptedEndpoint[1]
    const port = attemptedEndpoint[2]
    const path = attemptedEndpoint[3]
    return `http://${ip}:${port}/${path}`
  }
  return null
}

export const handlers = [
  rest.post('*/admin/generate_link', async (req, res, ctx) => {
    const type = await req.json().then(v => v.type)
    if (type === 'signup') {
      const originalResponse = await ctx.fetch(req)
      const jsonObject = await originalResponse.json()
      const userId = jsonObject['id']
      const endpointToNotify = emailToEndpoint(jsonObject['email'])
      if (userId && endpointToNotify) {
        fetch(endpointToNotify, {
          method: 'POST',
          body: JSON.stringify({
            userId,
            type
          })
        })
      }
      return res(ctx.json(jsonObject))
    } else {
      return req.passthrough()
    }
  }),

  rest.post('https://api.mailgun.net/v3/*/messages', async (req, res, ctx) => {
    console.log('Caught email request in handler-defined method')
    const bodyToHolder = req.body as {to: string, subject: string, text: string, html: string, from: string}
    const { to } = bodyToHolder
    
    const targetEndpoint = emailToEndpoint(to)

    if(targetEndpoint !== null) {
      console.log('found endpoint: ', targetEndpoint)
      let body = undefined
      try {
        body = JSON.stringify(req.body)
      } catch (err) {
        console.log('err: ', err)
      }

      fetch(targetEndpoint, {
        method: 'POST',
        body,
        headers: req.headers
      })
    }
    return res(ctx.status(200))
  }),
]
```

```
// test file (reduced, probably missing function definitions etc)
const dummyServer = http.createServer()
test.beforeAll(() => {
  dummyServer.listen(3001)
})
test.afterAll(() => {
  dummyServer.close()
})

test('User can log-in', async ({page}) => {
  const requests: any[] = [];

  dummyServer.addListener('request', (req, res) => {
    parseRequestBody(req, (body) => {
      const jbody = JSON.parse(body)
      requests.push(jbody)
    })
    res.writeHead(200)
    return res.end()
  })

 // ...

  // could add more detail here, like `test_${ip}_${port}_${path}_${behaviourType}@blah.com`
  const emailValue = `test${randomVal}_${localIP}_${port}_${path}_user@specificemailaddress.com`
  await emailBox.fill(emailValue);

  const submitButton = page.locator('role=button', { hasText: "REGISTER" })
  await expect(submitButton).toBeVisible();

  await submitButton.click()
  // assert against the contents of requests[0], requests[1] etc
```

Despite making this work, it's not ideal, and some IPC-type arrangement would be preferable. Do you think if I manipulated the next scripts to use code like that in the snippets for https://github.com/mswjs/interceptors/pull/129, then I'd have any chance of making that code work (I guess this would require working out how the non-child-process was a non-child, and trying to change that?)

If that's still a direction that you're thinking about, and I still have any mental composition after finishing this project for the time being (mid November), then I can try digging into the Next side or working out if a handling server can be spun up more easily (or an interface used to simplify what I'm doing above)? However, as is probably obvious, I know basically nothing about the majority of the required technology, so no promises of success (even assuming attempts).
Adding a link here because I also struggled to find this one again: https://github.com/vercel/next.js/discussions/26928
